clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -10908,6 +10908,9 @@ def err_sycl_restrict : Error<
   "|use a const static or global variable that is neither zero-initialized "
   "nor constant-initialized"
   "}0">;
+def warn_array_num_elements_exceeded
```

Comment:
Why is this only a warning?  Arrays larger won't work, right?   Why not make it an error?

Reply:
According to SPIR-V specification Section [2.17.Universal-Limits](https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.html#_a_id_limits_a_universal_limits) :

> 2.17. Universal Limits
> These quantities are minimum limits for all implementations and validators. Implementations are allowed to support larger quantities. **Specific APIs may impose larger minimums.**

This means that a SPIR-V module **can** have quantities **exceeding the minimum limits** (65,535 WC). It is not exactly an error for a SPIR-V module to have arrays larger than the minimum limit.

On the other hand it means the client API (device BE compilers) must be ready to execute such SPIR-V modules (which our implementation doesn't support currently)
Restricting this is probably not conformant with SYCL spec, but it is important to inform user about limitation of our implementation. Hence emitting this as a warning instead of an "error".

Reply:
If our implementation does not allow it to be executed correctly, we should issue an error.
Also, if this is a spir-v limitation, the error should be issued only for that target.

Reply:
A SPIR-V instruction **can not** have WC > 65535, because according to the spec we have only 16 bit to encode the WC. 65535 is the biggest integer number that can be expressed with 16 bit. An instruction with WC > 65535 can't be represented correctly in SPIR-V.

This WC limitation is not about our (or any other) implementation. It's SPIR-V fundamental limitation. So I think it must be an error, not a warning.

At the same time, instructions exceeding _other_ limits from 2.17 are possible and for them we should emit warnings.

Reply:
>It's SPIR-V fundamental limitation.
This wasn't obvious from the spec's description above. Thanks for the clarification. 
Thanks Prem for bringing this up.

Reply:
@keryell is asking the same as what I suggested above.  "if this is a spir-v limitation, the error should be issued only for that target."

Reply:
I guess I am not sure what you mean by target here.
My understanding is device code gets compiled into a device-agnostic form ( SPIRV), that can run on any compatible device.
This device-agnostic code(SPIRV) gets compiled into a device-specific form at runtime ( in the case of JIT). That is why, I added this check inside the `if(getLangOpts().SYCLIsDevice)` block for generic SPIR-V compilation(JIT)
When you say "error should be issued only for that target", are your referring to AOT compilation?  And what targets?

Reply:
> I guess I am not sure what you mean by target here.
> My understanding is device code gets compiled into a device-agnostic form ( SPIRV), that can run on any compatible device.

That is my point: what if the device is not compatible with SPIR-V?
An obvious example in DPC++ is the CUDA PTX device target.
So this error message should be emitted only for SPIR-V device code.
All this is based on Clang/LLVM, so the real device-agnostic forms are either Clang AST or LLVM IR. :-)

> This device-agnostic code(SPIRV) gets compiled into a device-specific form at runtime ( in the case of JIT). That is why, I added this check inside the `if(getLangOpts().SYCLIsDevice)` block for generic SPIR-V compilation(JIT)
> When you say "error should be issued only for that target", are your referring to AOT compilation? And what targets?

Ask your FPGA colleagues what they think about JIT... ;-)

Reply:
> what if the device is not compatible with SPIR-V?
> So this error message should be emitted only for SPIR-V device code.

Ok. That makes sense.

Reply:
> That is my point: what if the device is not compatible with SPIR-V?
> An obvious example in DPC++ is the CUDA PTX device target.
> So this error message should be emitted only for SPIR-V device code.

Fair enough. Can we distinguish cases when we are generating device code in SPIR-V format and cases when we use other format(s)?